vpr/src/pack/cluster_legalizer.h

```diff
@@ -0,0 +1,478 @@
+     * into different clusters. Should call the compress() method after destroying
+     * one or more clusters.
+     *
+     *  @param cluster_id       The ID of the cluster to destroy.
+     */
+    void destroy_cluster(LegalizationClusterId cluster_id);
+
+    /*
+     * @brief Compress the internal storage of clusters. Should be called after
+     *        a cluster is destroyed.
+     *
+     * Similar to the Netlist compress method. Will invalidate all Legalization
+     * Cluster IDs.
+     *
+     * This method can be quite expensive, so it is a good idea to batch many
+     * cluster destructions and then compress at the end.
+     */
+    void compress();
+
+    /*
+     * @brief A list of all cluster IDs in the legalizer.
+     *
+     * If the legalizer has been compressed (or no clusters have been destroyed)
+     * then all cluster IDs in this list will be valid and represent a non-empty
+     * legalization cluster.
+     */
+    cluster_range clusters() const {
+        return vtr::make_range(legalization_cluster_ids_.begin(), legalization_cluster_ids_.end());
+    }
+
+    /*
+     * @brief Check that the given cluster is fully legal.
+     *
+     * This method runs an intra_lb_route on the given cluster. This ignores
+     * the cluster legalization strategy set by the user. This method will not
+     * correct the problematic molecules, it will only return true if the
+     * cluster is legal and false if it is not.
+     *
+     *  @param cluster_id       The ID of the cluster to fully legalize.
+     *
+     *  @return                 True if the cluster is legal, false otherwise.
+     */
+    bool check_cluster_legality(LegalizationClusterId cluster_id);
+
+    /*
+     * @brief Cleans the cluster of unnessary data, reducing the memory footprint.
+     *
+     * After this function is called, no more molecules can be added to the
+     * cluster. This method will ensure that the cluster has enough information
+     * to generate a clustered netlist from the legalized clusters.
+     *
+     * Specifically, this frees the pb stats (which is used by the clusterer
+     * to compute the gain) and the router data of the cluster.
+     *
+     * TODO: The pb stats should really not be calculated or stored in the
+     *       cluster legalizer.
+     *
+     *  @param cluster_id   The ID of the cluster to clean.
+     */
+    void clean_cluster(LegalizationClusterId cluster_id);
+
+    /*
+     * @brief Verify that all atoms have been clustered into a cluster.
+     *
+     * This will not verify if all the clusters are fully legal.
+     */
+    void verify();
+
+    /*
+     * @brief Finalize the clustering.
+     *
+     * Before generating a Clustered Netlist, each cluster needs to allocate and
+     * load a pb_route. This method will generate a pb_route for each cluster
+     * and store it into the clusters' pb.
+     */
+    void finalize();
```

Comment:
Yes. This call is ALWAYS needed if you want to generate a clustered netlist from the cluster legalizer. (technically you do not have to if you called "clean" on all clusters; however, this just becomes a NOOP in that case. It would be too complicated to mention that and not worth it at all).

I added more to the description to make it clear that this is required to generate a clustered netlist.